### ISSUE TITLE
Thoughts on database management/migrations

### ISSUE THREAD

**Issue by LukasKalbertodt** [elan-ev/tobira#77]
#69 made me think about how we want to handle database management in general. I researched and thought a bit about this topic. These are my thoughts/conclusions.

--- 

Some notes (hopefully there is no disagreement on those?):

- Releasing a new version of Tobira should include some easy way to migrate the old database to the new schema. Otherwise updating is terribly annoying for sys admins.
- I want commands for DB operations, e.g. `tobira db setup`, `tobira db reset` and `tobira db run script.sql`.
- For convenience, just starting `tobira` should do some setup/migrations automatically, as long as it's clear what to do. In case of ambiguity, Tobira shouldn't guess.
- It will regularly happen that we (devs) switch between versions requiring different DB schemas (e.g. git checkout to review a PR, ...). It shouldn't be too annoying to get the database into the correct state. Notably, for development, the actual data in the database isn't important most of the time.
- I would prefer the scripts to be written in actual `.sql` files and not Rust strings. We can still embed those files into the application (`include_str!`).

## Suggestion: `down.sql`

In addition to the obviously needed "forwards" migration (`up.sql`), I think it would be beneficial to also include scripts to roll back one migration. 

## Suggestion: store migrations in DB

We would add a table `migrations` (or so) which would contain all migrations leading to the current state of the database. All migrations should have some unique ID. Running a migration would add a row to this table, undoing a migration would remove the row.

Furthermore, I think it would be useful for development to also store the raw `up.sql` and `down.sql` in a `text` column:

Saving `down.sql` allows us to roll back a migration even if the script is not currently available in the repository. That is useful if you switch from one feature branch (`foo`) to a different one (`bar`). To get the database in the state expected by `bar`, we have to roll back the migration of `foo` and then run the migation of `bar`. But since `foo`'s `down.sql` is not in the working directory anymore, we need to store it in the DB. Of course, you can always manually undo `foo`'s migration before switching to the new branch, but we will likely forget that most of the time and get annoyed by how inconvenient that is.

Saving `up.sql` allows us to compare the `up.sql` in the working directory/file system with the one in the database. If they differ (which will happen a lot during development), Tobira could easily undo the old migration (with the `down.sql` in the database) and then run the new `up.sql` from the file system.

### Solution A: one migration per version

One migration script per version (e.g. `v1.0.sql`, `v1.1.sql` and `next.sql`). We would only modify the latest (e.g. `next.sql`) during development. Releasing a new version means `mv next.sql 1.2.sql && touch next.sql`. The ID for the migrations in the `migration` table would simply be the version. That way, Tobira knows what migrations to run when updating to a new version.

- Advantage: works great for production users as they can easily upgrade to a new version
- Advantage: having all schema changes between two versions in one file helps on some situations
- Advantage: migration potentially faster as there are fewer scripts
- Disadvantage: for developers, switching between branches means to completely undo all database changes since the last released version of Tobira and then redoing them.

### Solution B: one migration per "feature" (commit/PR)

Instead of continuously changing one script until release, we could also have a separate migration script per "feature". That roughly corresponds to one migration script per PR that requires to modify the database schema. This of course results in lot more migration scripts, but also probably not ungodly many (adjust the DB schema is not too common). As ID for migrations, we would probably either want to use the current timestamp or some random short string. This is mainly to ensure that two feature branches developed in parallel will have different IDs for their migrations.

- Still works fine for production users, although it might be a bit slower than solution A, simply because there are more migrations to run.
- Advantage: more atomic migrations could make error handling/debugging easier
- Advantage: for development, switching between branches would only require undoing a small migration instead of basically resetting the whole database.
- Disadvantage: in solution (a) we could merge changes easily, whereas here, we might end up with migrations that undo part of previous migrations. Maybe. 

I at least know of one library which encourages this style of migration management.

**Comment by JulianKniephoff** [comment]
Just some perspective on how Rails, for example, handles the "multiple branches problem": Every time you run a migration, in addition to it being applied to the DB, a separate representation of the entire result schema is updated. (In the case of Rails, this is Ruby code with some additional magic, but in the simplest case, that could just be an SQL-dump from the DB). This representation is tracked in version control and can be used to recreate the database from scratch. That's often much simpler, and also faster, than to (be it manually or automatically) run the appropriate up- and down-migrations. It can of course also be used to quickly initialize production databases.

In addition to the schema, Rails also has the concept of "seed data", which we probably want in some form or another. This could for example be the root realm (although that could also be part of the schema; the boundaries here are of course blurry), or a bunch of test data.

Also note that in Rails, all of the above is Ruby code, not plain SQL. Just for the record I would like to note again that sometimes, and I personally think that it's more often than one would like to think, SQL is not enough or just too clunky to express everything a migration needs to do.

More comments after a more thorough read. :sweat_smile:

**Comment by lkiesow** [comment]
Some random thoughts and discussion points:

- +1 for getting SQL out of the code
- Do we really need the down script? Database snapshots are a very simple solution. Getting down may mean way more data and could mean a lot of work.
- If we have versioning, why do things manually? Detect versions and upgrade to the latest on startup.
- One version per change makes things easier and reduces potential errors. There won't be so many when we reach a production state anyway. We can introduce a breaking change before reaching a stable state anyway.
- Versioning the whole schema will not work. This might change based on database versions.

**Comment by LukasKalbertodt** [comment]
We discussed this in person and had a few more thoughts/conclusions:

- Writing `down.sql` is simple in the beginning, but once we actually migrate data, it's non trivial to write those. We we probably don't want those (especially if we can get most of the benefits in another way).
- This "another way" is probably just something like `docker commit` or other docker volumes tricks. Or really just `tobira db reset`. We also concluded that the situation with two feature branches changing different things about the database will probably be rather rare. 
- Consequently, storing the up/down script in the DB doesn't really make sense.
- We also agreed that we tend towards atomic migrations but we also don't mind manually combining a bunch of migration files at some point to reduce the number of migrations. 

I'm still unclear about if it's sufficient to store a single "version" in the database or whether we want to list all migrations performed. I guess we can require all migrations to be applied in a specific order, so in that sense, the newest applied migration version seems to be sufficient (as it implies all previous migrations were applied as well). But it might still be worth it to store some other data per migration? A hash of the migration script to make sure nothing fishy is going on? Dunno
